### PR TITLE
Introduce import engine and naming system

### LINKED ISSUE #87: Add identity system for symbols

### LINKED ISSUE #88: Handle environments imports

### PR DESCRIPTION
**Pull request by Override-6**
The branch `analyzer/identity-nd-imports` introduces a way to identify symbols by their names, and import/export them. To achieve these features, some significant changes were made.

## The `Name` structure
The `Name` structure will be an important part of the analyzer as it references a symbol, a module or an environment.
The structure is divided in a path prefix and a name string. The prefixed path is the path where the symbol is contained, and the name is the symbol's simple name.
A name's string representation is of form `path::to::symbol` where `path::to` is the prefix path and `symbol` is the Name's simple name.
A name cannot be empty.

## Introduce `ContextExports`
The `ContextExports` trait must be implemented by each Environment's contexts. It's used to search for exported symbols of a specific context.
Currently the two implementations of the ContextExports trait is the TypeContext (as it's the only context of the environment) and the Environment, which is a specific implementation that is used to aggregate its contexts.

## Introduce `ModuleLayers`
The `ModuleLayers` is a simple tree that contains all the modules with their environments.

## Introduce `ImportEngine`
Each `Environment` have their ImportEngines which determines their dependencies with other environments within the same `ModuleLayers`. 
The ImportEngine can be used to lookup symbols and list imported symbols that were never looked up.

### `ImportEngine` and `ReadOnlyImportEngine`
Each Environments have their `ImportEngine` and can add imports into them. Their contexts (for example the TypeContext) have a Read-only view of the environment's ImportEngine that does not contains way to modify the imports in the ImportEngine. But can still lookup symbols and list unused imports

## Other changes 

* Removed type contexts from `TypeClass`.
* Removed DefinedType which is now replaced by `ParametrizedType`
* 
This pull request fixes: 
* fixes #88 
* fixes #87

### CODE REVIEW

**Review by syldium — CHANGES_REQUESTED** [review]

File: analyzer/src/name.rs [hunk 1]
```diff
@@ -0,0 +1,107 @@
+use std::fmt::{Display, Formatter};
+
+
+#[derive(Debug, PartialOrd, Ord, Clone, PartialEq, Eq, Hash, Default)]
```
[the remark this answers is not in the document]
Comment (Override-6) [reply analyzer/src/name.rs:4]: :+1:

File: analyzer/src/import_engine.rs [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+    //remove all the symbols childrens of given symbol fqn
+    fn remove_all_in(&mut self, fqn: Name) {
+        self.imported_symbols = HashMap::from_iter(self.imported_symbols
+            .iter()
+            .filter(|(_, v)| {
+                let import_fqn = v.env_fqn.appended(v.symbol_name.clone());
+                fqn == import_fqn
+            })
+            .map(|(k, v)| (k.clone(), v.clone())))
```
Comment (syldium) [inline analyzer/src/import_engine.rs:222]:
```suggestion
        self.imported_symbols = self.imported_symbols
            .drain()
            .filter(|(_, v)| {
                let import_fqn = v.env_fqn.appended(v.symbol_name.clone());
                fqn == import_fqn
            })
            .collect();
```

File: analyzer/src/name.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    }
+
+    pub fn child(&self, name: &str) -> Self {
+        let mut parts = self.parts.clone();
+        parts.push(name.to_string());
+        Self { parts }
+    }
+
+    pub fn tail(&self) -> Option<Self> {
+        if self.parts.len() == 1 {
```
Comment (syldium) [inline analyzer/src/name.rs:73]:
```suggestion
        if self.path().is_empty() {
```

File: analyzer/src/name.rs [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+            .parts
+            .clone()
+            .into_iter()
+            .skip(common_parts_len)
+            .collect();
+        Some(Name::from(parts))
+    }
+
+    pub fn path(&self) -> impl Iterator<Item = &String> {
+        self.parts.iter().take(self.parts.len() - 1)
```
Comment (syldium) [inline analyzer/src/name.rs:55]:
```suggestion
    pub fn path(&self) -> &[String] {
        self.parts.split_last().unwrap().1 //Names cannot be empty
```

**Review by syldium — CHANGES_REQUESTED** [review]

File: analyzer/src/types/context.rs [hunk 5]
```diff
@@ -98,49 +163,29 @@ impl TypeContext {
...
-    }
-
-    pub(crate) fn fork(ctx: Rc<RefCell<Self>>) -> TypeContext {
-        TypeContext {
-            dependencies: vec![ctx],
-            ..Default::default()
+            None => self
+                .imports
+                .use_element::<Rc<TypeClass>, Self>(&name)
+                .ok_or(format!("Unknown type {}", &name)),
```
Comment (syldium) [inline analyzer/src/types/context.rs]:
```suggestion
                .ok_or_else(|| format!("Unknown type {name}")),
```

File: analyzer/src/name.rs [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        Self { parts }
+    }
+}
+
+impl From<Vec<String>> for Name {
+    fn from(value: Vec<String>) -> Self {
+        if value.is_empty() {
+            panic!("empty vec input")
+        }
+        Self { parts: value }
```
Comment (syldium) [inline analyzer/src/name.rs]:
```suggestion
        assert!(!value.is_empty(), "cannot create a name from an empty vec");
        Self { parts: value }
```

File: analyzer/src/types/context.rs [hunk 7]
```diff
@@ -1,71 +1,139 @@
...
+            .filter(|tc| tc.visibility == Public)
+            .cloned()
+    }
+
+    fn list_exported_names(&self, symbol: Option<Name>) -> Vec<Name> {
+        let parts = symbol.map(Name::into_vec).unwrap_or_default();
+        self.classes
+            .iter()
+            .filter_map(|(k, v)| {
+                if k.parts().starts_with(parts.as_slice()) {
```
Comment (syldium) [inline analyzer/src/types/context.rs]:
```suggestion
                if k.path() == parts.as_slice() {
```

**Review by syldium — APPROVED** [review]
Get ready :rocket:

**Pull request merged**